backend/linux_input.c

```diff
@@ -0,0 +1,188 @@
+            tev.u.pointer.screen_x = tm->x;
+            tev.u.pointer.screen_y = tm->y;
+            tev.u.pointer.button = tm->btns;
+            twin_screen_dispatch(tm->screen, &tev);
+        } else if (ev->code == ABS_Y) {
+            tm->y = ev->value;
+            _mouse_check_bounds(tm);
+            tev.kind = TwinEventMotion;
+            tev.u.pointer.screen_x = tm->x;
+            tev.u.pointer.screen_y = tm->y;
+            tev.u.pointer.button = tm->btns;
+            twin_screen_dispatch(tm->screen, &tev);
+        }
+        break;
+    case EV_KEY:
+        if (ev->code == BTN_LEFT) {
+            tm->btns = ev->value > 0 ? 1 : 0;
+            tev.kind = tm->btns ? TwinEventButtonDown : TwinEventButtonUp;
+            tev.u.pointer.screen_x = tm->x;
+            tev.u.pointer.screen_y = tm->y;
+            tev.u.pointer.button = tm->btns;
+            twin_screen_dispatch(tm->screen, &tev);
+        }
+    }
+}
+
+static void *twin_linux_evdev_thread(void *arg)
+{
+    twin_linux_input_t *tm = arg;
+
+    /* Open all event devices */
+    char evdev_name[EVDEV_NAME_SIZE_MAX] = {0};
+    for (int i = 0; i < EVDEV_CNT_MAX; i++) {
+        snprintf(evdev_name, EVDEV_NAME_SIZE_MAX, "/dev/input/event%d", i);
+        int fd = open(evdev_name, O_RDWR | O_NONBLOCK);
+        if (fd > 0) {
+            evdev_fd[evdev_cnt] = fd;
+            evdev_cnt++;
+        }
+    }
+
+    /* Initialize pollfd array */
+    struct pollfd pfds[EVDEV_CNT_MAX];
+    for (int i = 0; i < evdev_cnt; i++) {
+        pfds[i].fd = evdev_fd[i];
+        pfds[i].events = POLLIN;
+    }
+
+    /* Event polling */
+    struct input_event ev;
+    while (1) {
+        /* Wait until any event is available */
+        if (poll(pfds, evdev_cnt, -1) <= 0)
+            continue;
+
+        /* Iterate through all file descriptors */
+        for (int i = 0; i < evdev_cnt; i++) {
+            /* Try reading events */
+            ssize_t n = read(pfds[i].fd, &ev, sizeof(ev));
+            if (n == sizeof(ev)) {
+                /* Handle events */
+                twin_linux_input_events(&ev, tm);
+            }
+        }
+    }
+
+    return NULL;
+}
+
+static bool dummy(int file, twin_file_op_t ops, void *closure)
+{
+    return true;
+}
+
+void *twin_linux_input_create(twin_screen_t *screen)
+{
+    /* Create object for handling Linux input system */
+    twin_linux_input_t *tm = calloc(1, sizeof(twin_linux_input_t));
+    if (!tm)
+        return NULL;
+
+    tm->screen = screen;
+
+    /* Centering the cursor position */
+    tm->x = screen->width / 2;
+    tm->y = screen->height / 2;
+
+#if 1
+    /* FIXME: Need to fix memory leak before removing this line, which is purely
+     * coincident */
+
+    /* Set file handler for reading input device file */
+    twin_set_file(dummy, tm->fd, TWIN_READ, tm);
+#endif
+
+    /* Start event handling thread */
+    if (pthread_create(&tm->evdev_thread, NULL, twin_linux_evdev_thread, tm)) {
```

Comment:
WIN (and by extension, Mado) was designed to operate in single-threaded environments, and it is not reasonable to assume POSIX threads are available everywhere. Be sure to note the use of POSIX threads in the git commit messages for the Linux framebuffer backend. Coroutines could potentially be used as an alternative.

Reply:
Reference  on Coroutine in C:
https://github.com/sysprog21/mado/issues/4#issuecomment-2370122526